src/arithmetic/string_funcs/string_funcs.c

```diff
@@ -167,6 +168,182 @@ SIValue AR_SUBSTRING(SIValue *argv, int argc, void *private_data) {
+	if(rv != ONIG_NORMAL) {
+		char s[ONIG_MAX_ERROR_MESSAGE_LEN];
+		onig_error_code_to_str((UChar* )s, rv, &einfo);
+		ErrorCtx_SetError("invalid regex, err=%s", s);
+		return SI_NullVal();
+	}
+	ASSERT(rv == ONIG_NORMAL);
+
+	_match_regex_scan_cb_args args = {
+		.list = &list,
+		.str = str
+	};
+
+	rv = onig_scan(regex, (const UChar *)str,
+			(const UChar *)(str + strlen(str)), region, ONIG_OPTION_DEFAULT, _match_regex_scan_cb, &args);
+	if(rv < 0) {
+		char s[ONIG_MAX_ERROR_MESSAGE_LEN];
+		onig_error_code_to_str((OnigUChar* )s, rv);
+		ErrorCtx_SetError("invalid regex, err=%s", s);
+		return SI_NullVal();
+	}
+
+	onig_free(regex);
+	onig_region_free(region, 1);
+
+	return list;
+}
+
+typedef struct {
+	char *res;
+	uint32_t res_len;
+	const char *str;
+	uint32_t str_ind;     		// current index in str
+	const char *replacement;
+	uint32_t replacement_len;
+} _replace_regex_scan_cb_args;
+
+static int _replace_regex_scan_cb(int n, int pos, OnigRegion *region, void *arg) {
+	_replace_regex_scan_cb_args *args = (_replace_regex_scan_cb_args *)arg;
+	const char *str = args->str;
+	SIValue subList = SIArray_New(region->num_regs);
+
+	int match_len = region->end[0] - region->beg[0];
+	int str_copy_len = region->beg[0] - args->str_ind;
+	args->res =
+	rm_realloc(args->res, (args->res_len + match_len + args->replacement_len + 1)*sizeof(char));
+
+	// copy the string between the last match and the current match
+	memcpy(args->res + args->res_len, str + args->str_ind, str_copy_len*sizeof(char));
+	args->str_ind = region->end[0];
+	args->res_len += str_copy_len;
+
+	// copy the replacement string
+	memcpy(args->res + args->res_len, args->replacement, args->replacement_len*sizeof(char));
+	args->res_len += args->replacement_len;
+
+	args->res[args->res_len] = '\0';
+
+	return 0;
+}
+
+// given a string and a regular expression,
+// return a string after replacing each regex match with a given replacement.
+// string.replaceRegEx(str, regex, replacement) → string
+SIValue AR_REPLACEREGEX(SIValue *argv, int argc, void *private_data) {
+	ASSERT(argc == 2 || argc == 3);
+	if(SI_TYPE(argv[0]) == T_NULL || SI_TYPE(argv[1]) == T_NULL) {
+		return SI_NullVal();
+	}
+
+	ASSERT(SI_TYPE(argv[0]) == T_STRING);
+	ASSERT(SI_TYPE(argv[1]) == T_STRING);
+
+	const char *str = argv[0].stringval;
+	const char *regex_str = argv[1].stringval;
+	char *replacement = "";
+
+	if(argc == 3) {
+		if(SI_TYPE(argv[2]) == T_NULL) {
+			return SI_NullVal();
+		}
+
+		ASSERT(SI_TYPE(argv[2]) == T_STRING);
+		replacement = argv[2].stringval;
+	}
+
+	int rv;
+	regex_t *regex;
+
+	OnigRegion *region = onig_region_new();
+	OnigErrorInfo einfo;
+
+	rv = onig_new(&regex, (const UChar *)regex_str, (const UChar *)(regex_str + strlen(regex_str)),
+			 ONIG_OPTION_DEFAULT, ONIG_ENCODING_UTF8, ONIG_SYNTAX_JAVA, &einfo);
+	if(rv != ONIG_NORMAL) {
+		char s[ONIG_MAX_ERROR_MESSAGE_LEN];
+		onig_error_code_to_str((UChar* )s, rv, &einfo);
+		ErrorCtx_SetError("invalid regex, err=%s", s);
```

Comment:
@vityafx It's ok to call onig_free on null ptr, since the function makes this check already.
When there is an error onig_new will set regex to NULL by itself.